C++ tests fail to build on macOS due to mkdtemp not declared

Building the C++ tests on macOS fails with:
```
tests/cpp/filesystem.cc:61:18: error: use of undeclared identifier 'mkdtemp'
```
Fix is to include `<unistd.h>` under `#if defined(__APPLE__)` in `tests/cpp/filesystem.cc`, which exposes `mkdtemp` on macOS.
Tested on macOS Apple Silicon (arm64).

--- a/tests/cpp/filesystem.cc
+++ b/tests/cpp/filesystem.cc
@@ -9,6 +9,10 @@
 
 #if !defined(xgboost_IS_WIN)
 
+#if defined(__APPLE__)
+#include <unistd.h>  // for mkdtemp on macOS
+#endif
+
 #include <cstdlib>  // for mkdtemp
 
 #include "../../src/common/error_msg.h"  // for SystemError

